components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java

```diff
@@ -0,0 +1,459 @@
+		this.useRecovery = useRecovery;
+	}
+
+	@Override
+	public boolean isUseRecovery() {
+		return useRecovery;
+	}
+
+	@Override
+	public void setDeadLetterUri(String deadLetterUri) {
+		this.deadLetterChannel = deadLetterUri;
+	}
+
+	@Override
+	public String getDeadLetterUri() {
+		return deadLetterChannel;
+	}
+
+	@Override
+	public void setMaximumRedeliveries(int maximumRedeliveries) {
+		this.maximumRedeliveries = maximumRedeliveries;
+	}
+
+	@Override
+	public int getMaximumRedeliveries() {
+		return maximumRedeliveries;
+	}
+
+	@Override
+	public Exchange get(CamelContext camelContext, String key) {
+		CompletableFuture<GetResponse> completableResponse = kvClient
+				.get(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()));
+		try {
+			GetResponse getResponse = completableResponse.get();
+			byte[] data = getResponse.getKvs().get(0).getValue().getBytes();
+			ByteArrayInputStream in = new ByteArrayInputStream(data);
+			ObjectInputStream is = new ObjectInputStream(in);
+			DefaultExchangeHolder holder = (DefaultExchangeHolder) is.readObject();
+			return unmarshallExchange(camelContext, holder);
+		} catch (InterruptedException | ExecutionException | IOException | ClassNotFoundException e) {
+			LOG.error(e.getMessage(), e);
+		}
+		return null;
+	}
+
+	public boolean isAllowSerializedHeaders() {
+		return allowSerializedHeaders;
+	}
+
+	public void setAllowSerializedHeaders(boolean allowSerializedHeaders) {
+		this.allowSerializedHeaders = allowSerializedHeaders;
+	}
+
+	@Override
+	public void remove(CamelContext camelContext, String key, Exchange exchange) {
+		DefaultExchangeHolder holder = DefaultExchangeHolder.marshal(exchange, true, allowSerializedHeaders);
+		if (optimistic) {
+			LOG.trace("Removing an exchange with ID {} for key {} in an optimistic manner.", exchange.getExchangeId(),
+					key);
+			CompletableFuture<DeleteResponse> completableDeleteResponse = kvClient
+					.delete(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()));
+			try {
+				DeleteResponse deleteResponse = completableDeleteResponse.get();
+				if (deleteResponse.getDeleted() == 0) {
+					LOG.error(
+							"Optimistic locking failed for exchange with key {}: kvClient.delete removed no Exchanges, while it's expected to remove one.",
+							key);
+					throw new OptimisticLockingException();
+				}
+			} catch (InterruptedException | ExecutionException e) {
+				LOG.error(e.getMessage(), e);
+			}
+			LOG.trace("Removed an exchange with ID {} for key {} in an optimistic manner.", exchange.getExchangeId(),
+					key);
+			if (useRecovery) {
+				LOG.trace(
+						"Putting an exchange with ID {} for key {} into a recoverable storage in an optimistic manner.",
+						exchange.getExchangeId(), key);
+				try {
+					ByteArrayOutputStream bos = new ByteArrayOutputStream();
+					ObjectOutputStream oos = new ObjectOutputStream(bos);
+					oos.writeObject(holder);
+					oos.flush();
```

Comment:
seems like serialize/deserialize of the exchange holder is quite common so you may think to extract the logic in some methods